solidity/contracts/TokenizedAccountNFT.sol

```diff
@@ -0,0 +1,122 @@
+     * of each token should be left in the TokenizedAccount's x/bank account.
+     *
+     * Any excess balances will accumulate here, and may be retrieved by the owner with
+     * the withdrawBalances(erc20s) method
+     *
+     * The ERC20s specified here have EVM addressses, querying the x/erc20 module will reveal their
+     * Cosmos Coin counterparts
+     *
+     * @notice this function is access controlled, only the owner or an approved msg.sender may call this function
+     */
+    function setThresholds(address[] calldata newErc20s, uint256[] calldata newAmounts) public virtual onlyOwnerOrApproved(AccountId) {
+        require(newErc20s.length == newAmounts.length, "threshold values must have the same length");
+        // Clear the thresholds before overwriting
+        delete thresholdErc20s;
+        delete thresholdAmounts;
+
+        for (uint i=0; i<newErc20s.length; i++) {
+            thresholdErc20s.push(newErc20s[i]);
+            thresholdAmounts.push(newAmounts[i]);
+        }
+        emit ThresholdsChanged(newErc20s, newAmounts);
+    }
+
+    /**
+     * @dev Withdraws ERC20 balances from this contract to the owner, throws on any erc20 transfer() failure
+     * Emits a {SuccessfulWithdrawal} event upon success.
+     *
+     * @notice This function is access controlled, only the owner or an approved msg.sender may call this function
+    */
+
+    // TODO: Should we even accept a destination, or just send to the owner of the AccountId token?
+    function withdrawBalances(address[] calldata erc20s, address destination) public virtual onlyOwnerOrApproved(AccountId) {
+        for (uint i=0; i<erc20s.length; i++) {
+            address erc20 = erc20s[i];
+            uint256 balance = IERC20(erc20).balanceOf(address(this));
+            if (balance > 0) {
+                bool result = IERC20(erc20).transfer(destination, balance);
+                require(result, "unsuccessful withdrawal");
+            }
+        }
+        emit SuccessfulWithdrawal(erc20s);
+    }
+
+    /**
+     * @dev Begins the TokenizedAccount recovery process, which must be detected by the x/microtx module.
+     * Emits a {TryRecover} event which must be detected by the x/microtx module.
+     * Expected behavior is that all tokens held in the x/bank account of the TokenizedAccount will be
+     * converted to ERC20s and transferred to the control of this contract. If the recovery is successful
+     * then the x/microtx module will append a {SuccessfulRecovery} event with the ERC20 addresses and amounts
+     * sent to this contract.
+     *
+     * After a successful recovery, use withdrawBalances(erc20s) to send the token balances to the owner account.
+     *
+     * @notice Due to the EVM <> Cosmos interactions, the tokens will not be transferred to this contract until
+     * the entire transaction is resolved. It will not be possible to recoverAccount() and immediately withdrawBalances()
+     * in the same transaction!
```

Comment:
shouldn't this be 'the same block' since hooks are called per block?

Reply:
No, it's at the end of a transaction (in the middle of a block). The flow looks like this:

1. Anything like an eth_sendTransaction is converted into a MsgEthereumTx, then executed as a Cosmos Tx.
2. In the msg_server, the Cosmos Tx is decoded back to an Ethereum transaction, then applied via geth common.
3. After the EVM execution, the PostTxProcessing hooks are called. x/microtx will be using one of those PostTxProcessing hooks to detect EVM log events and performing state changes + log changes.
4. Finally the logs (which may be modified by PostTxProcessing hooks) are emitted as a cosmos event